File: backend/python-api/app/service/user_service.py
from __future__ import annotations

from app.domain.errors import ERR_CONFLICT, ERR_INVALID_INPUT, DomainError
from app.domain.username import normalize_username, validate_username
from app.domain.user import User
from app.dto.user import UpdateProfileRequest
from app.i18n.loader import is_language_supported
from app.repository.user_repo import UserRepository


class UserService:
    def __init__(self, users: UserRepository):
        self._users = users

    async def get_profile(self, user_id: int) -> User:
        return await self._users.find_by_id(user_id)

    async def update_profile(self, user_id: int, req: UpdateProfileRequest, session) -> User:
        user = await self._users.find_by_id(user_id)

        if req.username is not None:
            username = normalize_username(req.username)
            err = validate_username(username)
            if err:
                raise DomainError(ERR_INVALID_INPUT, err)
            available = await self._users.is_username_available(username, exclude_id=user.id)
            if not available:
                raise DomainError(ERR_CONFLICT, "Username already taken")
            user.username = username

        if req.first_name is not None:
            user.first_name = req.first_name
        if req.last_name is not None:
            user.last_name = req.last_name
        if req.language is not None:
            if is_language_supported(req.language.lower()):
                user.language = req.language.lower()
        if req.country is not None:
            user.country = req.country.upper()[:10]
        if req.address is not None:
            user.address = req.address
        if req.phone is not None:
            user.phone = req.phone
        if req.two_fa_enabled is not None:
            user.two_fa_enabled = req.two_fa_enabled
            if not req.two_fa_enabled:
                user.two_fa_secret = None

        await self._users.update(user)
        await session.commit()
        return user

File: backend/go-api/internal/service/user_service.go
package service

import (
	"context"
	"log"
	"strings"

	"saas-starter/backend/go-api/internal/domain"
	"saas-starter/backend/go-api/internal/dto"
	"saas-starter/backend/go-api/internal/repository"
	"saas-starter/backend/go-api/pkg/i18n"
)

type UserService struct {
	users repository.UserRepository
}

func NewUserService(users repository.UserRepository) *UserService {
	return &UserService{users: users}
}

func (s *UserService) GetProfile(ctx context.Context, userID uint) (*domain.User, error) {
	return s.users.FindByID(ctx, userID)
}

func (s *UserService) UpdateProfile(ctx context.Context, userID uint, req dto.UpdateProfileRequest) (*domain.User, error) {
	user, err := s.users.FindByID(ctx, userID)
	if err != nil {
		return nil, err
	}

	if req.Username != "" {
		normalized := domain.NormalizeUsername(req.Username)
		if err := domain.ValidateUsername(normalized); err != nil {
			return nil, domain.NewError(domain.ErrInvalidInput, err.Error())
		}
		available, err := s.users.IsUsernameAvailable(ctx, normalized, user.ID)
		if err != nil {
			return nil, err
		}
		if !available {
			return nil, domain.NewError(domain.ErrConflict, "username already taken")
		}
		user.Username = normalized
	}

	if req.FirstName != "" {
		user.FirstName = req.FirstName
	}
	if req.LastName != "" {
		user.LastName = req.LastName
	}
	if req.Language != "" {
		langCode := req.Language
		if len(langCode) > 2 {
			langCode = langCode[:2]
		}
		if i18n.IsLanguageSupported(langCode) {
			user.Language = langCode
		}
	}
	if req.Country != "" {
		country := strings.ToUpper(req.Country)
		if len(country) > 10 {
			country = country[:10]
		}
		user.Country = country
	}
	if req.Address != "" {
		user.Address = req.Address
	}
	if req.Phone != "" {
		user.Phone = req.Phone
	}

	if req.TwoFAEnabled != nil {
		user.TwoFAEnabled = *req.TwoFAEnabled
		if !*req.TwoFAEnabled && user.TwoFASecret != "" {
			user.TwoFASecret = ""
			log.Printf("Cleared TOTP secret for user %d (2FA disabled)", user.ID)
		}
	}

	if err := s.users.Update(ctx, user); err != nil {
		return nil, err
	}

	return user, nil
}
